File: csharp/Pokemon/Pokemon/main/Gengar.cs
using System;
using System.Collections.Generic;
using System.Runtime.ConstrainedExecution;
using System.Text;

namespace Pokemon.main
{
    internal class Gengar : PokemonCharacter
    {
        private readonly int HIT_POINTS = 100;
        private readonly double DEFENSE_MULTIPLIER = 0.4;

        public Gengar()
        {
            Psychic psychic = new Psychic();
            Thunderpunch thunderpunch = new Thunderpunch();
            setType("Ghost");
            setName("Gengar");
            setHasEvolution(true);
            setSecondAttack(psychic);
            setMainAttack(thunderpunch);
            setHitPoints(HIT_POINTS);
            setDefenseMultiplier(DEFENSE_MULTIPLIER);
            setMainAttackDamage(thunderpunch.getAttackDamage());
            setSecondAttackDamage(psychic.getAttackDamage());
        }

        override
        public string evolve()
        {
            return null;
        }

        override
        public string defend(int attack)
        {
            int damage;

            damage = (int)(attack * getDefenseMultiplier());
            int newHP = getHitPoints() - damage;

            String defendMessage = new String("Defending attack, damage caused is "
                                              + damage + " new HP is " + newHP);

            setHitPoints(newHP);
            return defendMessage;

        }

        override
        public string secondAttack()
        {

            String attackMessage = new String("Attacking opponent with "
                                              + getSecondAttack()
                                              + " causing a damage of " + getSecondAttackDamage());
            return attackMessage;


        }

        override
        public string mainAttack()
        {
            String attackMessage = new String("Attacking opponent with "
                                              + getMainAttack()
                                              + " causing a damage of " + getMainAttackDamage());
            return attackMessage;

        }

        override
        public void setNewAttack(int attack, int attackDamage,
             IAttack newAttack)
        {
            if (attack == 1)
            {
                setMainAttack(newAttack);
                setMainAttackDamage(attackDamage);
            }
            else
            {
                setSecondAttack(newAttack);
                setSecondAttackDamage(attackDamage);
            }
        }

        public override string getAttack()
        {
            throw new NotImplementedException();
        }
    }
}

File: python/Gengar.py
from PokemonCharacter import PokemonCharacter
from ShadowPunch import ShadowPunch
from Hex import Hex


class Gengar(PokemonCharacter):
    HIT_POINTS = 58
    DEFENSE_MULTIPLIER = 0.5
    MAIN_ATTACK_DAMAGE = 4
    SECOND_ATTACK_DAMAGE = 12

    def __init__(self):
        self.setType("venom, phantom")
        self.setName("Gengar")
        self.setHasEvolution(True)
        self.setSecondAttack(ShadowPunch())
        self.setMainAttack(Hex())
        self.setHitPoints(HIT_POINTS)
        self.setDefenseMultiplier(DEFENSE_MULTIPLIER)
        self.setMainAttackDamage(MAIN_ATTACK_DAMAGE)
        self.setSecondAttackDamage(SECOND_ATTACK_DAMAGE)

    def evolve(self):
        return "Null"

    def defend(self, attack):
        int damage = attack * self.getDefenseMultiplier()
        newHP = self.getHitPoints() - damage
        deffendMessage = 'Defending attack, damage caused is ' + damage +
        ' new HP is ' + newHP
        return deffendMessage

    def secondAttack(self):
        attackMessage = 'Attacking opponent with ' + self.getSecondAttack() +
        ' causing a damage of ' + self.getSecondAttackDamage()
        return attackMessage

    def secondAttack(self):
        attackMessage = 'Attacking opponent with ' + self.getMainAttack() +
        ' causing a damage of ' + self.getMainAttackDamage()
        return attackMessage

    def setNewAttack(self, attack, attackDamage, newAttack):
        if (attack == 1):
            self.setMainAttack(newAttack)
            self.setMainAttackDamage(attackDamage)
        else:
            self.setSecondAttack(newAttack)
            self.setSecondAttackDamage(attackDamage)
